### DISCUSSION TITLE
6502 Processor

### DISCUSSION THREAD

**Discussion by jllancee** [hneemann/Digital#1152] [Q&A]
Hi, I am an undergraduate student who is working on making a model of 6502 processor in Digital software. This is based on a 6502 Logism file 6502.circ (see google drive link). I am getting an error when starting the simulation in Digital (photo attached).  Files are inside the google drive link. Does anyone have a workaround for this? Appreciate your help. Thanks!
[[https://drive.google.com/drive/folders/1Qx7kxJC0kgGlBitxH1MQk2_apWUsASB1?usp=sharing])](https://drive.google.com/drive/folders/1Qx7kxJC0kgGlBitxH1MQk2_apWUsASB1?usp=share_link)
![image](https://user-images.githubusercontent.com/97872630/234035500-0aa51223-5294-45b6-aec6-4eeb16f989a4.png)
![image](https://user-images.githubusercontent.com/97872630/234036498-6ab5839a-6d7b-4d8f-a452-a148deb6694b.png)

**Comment by mengstr** [comment]
I tried to run the main file and made a note of the statues of the wires leading into the ALU module. I then opened the ALU and ran it and setup the inputs the same as what the main was putting into it. It turns out that the `Alu_add` and `ALU_AC` modules both tries to output data on the SB_out bus.  There are so many control signals that I can't wrap my head around everything, so I can't help you further. But at least you have something to work from.
<img width="1323" alt="6502" src="https://user-images.githubusercontent.com/325326/234087933-84750c89-8bef-4655-83fa-75c8b72b0c51.png">

**Reply by jllancee** [reply]
Thanks!

**Comment by logoliv** [comment]
@jllancee Could you please share your project on Github if it works ? I think many people would like to simulate a 6502 with Digital 😄

**Reply by completely-unstable** [reply]
i have one that can at least run fibonocci:

https://github.com/user-attachments/assets/4c819c9c-4023-4e81-9b25-5e627194571f

and it runs actual assembled machine code:

![Untitled 18](https://github.com/user-attachments/assets/0f614ccd-6f53-4e5f-8b7a-9d01f38e8d11)

but im still working on getting all the instructions implemented correctly. also how would i upload to github, do i just need to include every dependency?

**Comment by SirClickALot** [comment]
I strongly second that!
@jllancee please do share it, it would be very useful indeed.

**Comment by shawty** [comment]
I started writing a 6502 & 6522 simulation for Digital using the Java Components extension point, log time ago... then work got in the way, and nothing ever got finished.

Just had a quick look, I thought I'd actually uploaded them to my github page, but if I remember correctly, I did actually raise something as a issue/help request on here in digital's early days.

Just had a poke about to try and find the thread, it turns out I had it in the "Custom Components" repo.

From the quick read I had, I think I didn't get much further than starting to implement the 6522.

You can find the thread here

https://github.com/hneemann/digitalCustomComponents/issues/3

With zip files containing the Java code still intact and downloadable.

**Comment by completely-unstable** [comment]
@logoliv @SirClickALot https://github.com/completely-unstable/6502-Digital

**Reply by shawty** [reply]
@completely-unstable "**this is not a 1:1 implementation, i designed everything via my own intuition and understanding, so there are nuances in things like how many clock cycles an instruction might take**" 

THIS is the book you need:

https://archive.org/details/Programming_the_6502_OCR

It has one page for every instruction (One screen shot page below)

![image](https://github.com/user-attachments/assets/2a573f62-f9fe-44dd-bf91-b54741a7dade)

Each page shows the logic function (Red Arrow), the Binary Format (Orange Underline), Op-Code (Green Arrow), Number of Bytes (Blue Arrow), Number of Cycles (Purple Arrow) and Flags Affected (Turquoise Underline)

I have an actual paper copy of the book (Bought it back in the 1980's, First Edition) when I first started to learn 6502 code on my BBC Model B micro computer.  :-)

Today I do also have a legitimately purchased eBook version on my kindle too.

Shawty

**Reply by completely-unstable** [reply]
cool, ill check it out. i would really like to know what exactly are the control signals that get set each cycle, ive found the # of cycles but dont really know what happens during them.

**Reply by shawty** [reply]
@completely-unstable in most cases the 6502 is really a very simple step wise CPU.

If an instruction states it takes 2 clock cycles, then your most likely looking at an instruction decode followed by a load, 3 cycles usually an ins decode followed by a 2 byte operation of some kind, the only time it starts to get a bit tricky is with Zero page, and long addresses, Rodney Zaks book that I mentioned above will take you through everything like that, there is a wealth of information in the book.

**Reply by shawty** [reply]
@completely-unstable here we go.  I knew I had this somewhere :-)

[6502 Schematic.pdf](https://github.com/user-attachments/files/17202377/6502.Schematic.pdf)

Enjoy.....

**Reply by hneemann** [reply]
@shawty 
This schematic is amazing! Thank you very much for sharing!

**Reply by shawty** [reply]
@hneemann your more than welcome :-)   I can't take credit for drawing it, but I have had it for a long time, I don't remember where it originally came from, but I suspect it was back in the day as a teenage programmer, writing off to electronics companies begging them for technical info when learning how to program the 6502 back in the 1980's.

That PDF is just a scan, somewhere (I do not know where) I will have the original document.
